### DISCUSSION TITLE
Can Dependency Injection be applied based on traits?

### DISCUSSION THREAD

**Discussion by myyrakle** [tokio-rs/axum#2047] [Q&A] [answered]
I tried to implement DI using Extension and Layer.

First, I defined an interface trait for Service and prepared two implementations.
![image](https://github.com/tokio-rs/axum/assets/16988115/af6112cb-bffe-4cf8-8123-5f1e2a70bb6c)

Second, I registered one of the implementations with the layer.
![image](https://github.com/tokio-rs/axum/assets/16988115/5dd829d4-bc7e-46ee-b76c-c912c2d9992b)

Thirdly, I tried importing a service based on a trait.
![image](https://github.com/tokio-rs/axum/assets/16988115/372dd3d2-87ee-4b49-89bc-55f91857ae47)

But it failed.
![image](https://github.com/tokio-rs/axum/assets/16988115/879a9c9a-9202-4cfe-b3f1-9b6e7a6457c8)

Perhaps now, axum's parameter DI seems to perform injection only for exactly matched types.

Is there a workaround for this or a recommended workaround?

I would like to be able to abstract the Service module and separate test versions to make unit testing easier...

**Comment by caenguidanos** [comment, marked as answer]
Yes, it's possible.

Try with `State` extractor to ensure correct type checking (it's recommendes from Axum docs).

First, build the `Dyn<>` type: https://github.com/caenguidanos/http-api-rs/blob/main/src/contexts/ecommerce/backoffice/domain/product/repository.rs.

Then, define the concrete implementation: https://github.com/caenguidanos/http-api-rs/blob/main/src/contexts/ecommerce/common/infrastructure/dependency_container.rs.

Finally, inject `with_state` on Router https://github.com/caenguidanos/http-api-rs/blob/main/src/contexts/ecommerce/backoffice/infrastructure/controller.rs and extract from handler https://github.com/caenguidanos/http-api-rs/blob/main/src/contexts/ecommerce/backoffice/infrastructure/http/get_products.rs.

**Reply by myyrakle** [reply]
Thanks for your kind reply!
